### PR TITLE
[RELAND] Added optimizers based on multi tensor apply

### PR DESCRIPTION
**Pull request by izdeby**
Original PR https://github.com/pytorch/pytorch/pull/45299.  The present PR fixes minor bugs that caused revert.

Adding a new namespace `torch.optim._multi_tensor` with a bunch of updated optimizers. Those optimizers are using _foreach APIs which improve performance significantly. 

### Tests
- updated existing tests to use both optimizers
- added `test_multi_tensor_optimizers` test to verify correctness. 

### Perf results

**Adam**
timeit: 42.69 ms --> 10.16 ms
autorange: 41.96 ms --> 10.28 ms

**AdamW**
timeit: 51.38 ms --> 15.63 ms
autorange: 50.82 ms --> 16.07 ms

**SGD**
timeit: 6.28 ms --> 4.40 ms
autorange: 6.13 ms --> 4.73 ms

**RMSprop**
timeit: 28.63 ms --> 5.89 ms
autorange: 28.27 ms -->  5.76 ms

**Rprop**
timeit: 213.30 --> 178.42
autorange: 212.03 --> 178.03

**ASGD**
timeit: 21.67 --> 9.33
autorange: 21.64 --> 9.27

**Adamax**
timeit: 55.60 --> 48.29
autorange: 55.22 -> 49.13

**Rerf Script used**

```
import torch
import time
import torch.optim as optim
from torch.autograd import Variable
from torch.optim.lr_scheduler import ExponentialLR, ReduceLROnPlateau, StepLR
import torch.nn as nn
import time
import torchvision
import torch.utils._benchmark as benchmark_utils

device = "cuda"
model = torchvision.models.resnet.resnet101(pretrained=True).to(device)
targets = torch.randint(0, 1000, (100, 100), device=device)
criterion = nn.CrossEntropyLoss()

optimizer = optim.SGD(model.parameters(), lr=1e-3) # <----------------------- optimizer. 
                                                          # would compare optim.SGD vs optim._multi_tensor.SGD
running_loss = 0.0
target = torch.empty(128, dtype=torch.long, device=device).random_(5)

optimizer.zero_grad()
inputs = torch.rand(128, 3, 100, 100, device=device , requires_grad=True)
outputs = model(inputs)
loss = criterion(outputs, target) 
loss.backward()
optimizer.step()
running_loss += loss.item()

def main():
    timer = benchmark_utils.Timer(
        stmt="optimizer.step()",
        globals=globals(),
        label="str(optimizer)",
    )

    for i in range(1):
        print(f"Run: {i}\n{'-' * 40}")
        print(f"timeit:\n{timer.timeit(1000)}\n")
        print(f"autorange:\n{timer.blocked_autorange()}\n\n")


if __name__ == "__main__":
    main()
```

### CODE REVIEW

File: torch/optim/_multi_tensor/sgd.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            if grads == []:
+                return loss
+
+            if weight_decay != 0:
+                grads = torch._foreach_add(grads, params_with_grad, alpha=weight_decay)
+
+            if momentum != 0:
+                bufs = []
+                for i in range(len(states)):
+                    if 'momentum_buffer' not in states:
```
Comment (mcarilli) [inline torch/optim/_multi_tensor/sgd.py:120]: Should this be `states[i]`?

Comment (izdeby) [reply torch/optim/_multi_tensor/sgd.py]: Done

File: torch/optim/_multi_tensor/sgd.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                grads = torch._foreach_add(grads, params_with_grad, alpha=weight_decay)
+
+            if momentum != 0:
+                bufs = []
+                for i in range(len(states)):
+                    if 'momentum_buffer' not in states[i]:
+                        buf = states[i]['momentum_buffer'] = torch.clone(grads[i]).detach()
+                    else:
+                        buf = states[i]['momentum_buffer']
+                        buf.mul_(momentum).add_(grads[i], alpha=1 - dampening)
```
Comment (mcarilli) [inline torch/optim/_multi_tensor/sgd.py:124]: This should also be repackaged as an MTA calculation, it's dominating the runtime in my tests (see picture).

File: torch/optim/_multi_tensor/sgd.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                if p.grad is not None:
+                    grads.append(p.grad)
+                    params_with_grad.append(p)
+                    states.append(self.state[p])
+
+                    if p.grad.is_sparse:
+                        has_sparse_grad = True
+
+                        if momentum != 0: 
+                            raise RuntimeError('SGD does not support momentum for sparse gradients')
```
Comment (mcarilli) [inline torch/optim/_multi_tensor/sgd.py:109]: The CPU overhead of this region is substantial (corresponds to the `"cpu packaging"` nvtx range of the attached profile).

Profile is for resnet50+native amp.  There's an amp-related sync (CPU waits for the GPU) just before SGD.step(), so the CPU latency is fully exposed to the GPU and causes a bubble on the GPU timeline.  For profiling purposes that's a good thing, because it clearly shows how much CPU overhead we have.  In a non-amp run, the overhead might not cause a GPU bubble because there's no sync, so the CPU might have enough time to get the packaging done and the optimizer kernels enqueued while the GPU is still processing earlier work.  (In amp-enabled runs, in principle we can similarly reduce the risk of a GPU bubble with "sync-free dynamic loss scaling," an extra layer of complexity we should consider soon, but not for 1.7.)

With or without an amp sync (and whether or not it causes a bubble on the GPU) there's a big chunk of CPU overhead here.  Some overhead is inevitable, but if there's time tomorrow, we should consider if any Python tricks can moderately reduce the overhead.
mta_sgd_step

File: torch/optim/_multi_tensor/adadelta.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                        state['acc_delta'] = torch.zeros_like(p, memory_format=torch.preserve_format)
+
+                    square_avgs.append(state['square_avg'])
+                    acc_deltas.append(state['acc_delta'])
+
+                    state['step'] += 1
+                    states.append(state)
+
+            if group['weight_decay'] != 0:
+                torch._foreach_add_(grads, params_with_grad, alpha=group['weight_decay'])
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/adadelta.py:81]: can you add a test that covers this case with `weight_decay`?

File: torch/optim/_multi_tensor/adamax.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                        state['exp_inf'] = torch.zeros_like(p, memory_format=torch.preserve_format)
+
+                    exp_avgs.append(state['exp_avg'])
+                    exp_infs.append(state['exp_inf'])
+
+                    state['step'] += 1
+                    states.append(state)
+
+            if group['weight_decay'] != 0:
+                torch._foreach_add_(grads, params_with_grad, alpha=group['weight_decay'])
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/adamax.py:85]: same

File: torch/optim/_multi_tensor/adamw.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                    state['exp_avg_sq'] = torch.zeros_like(p, memory_format=torch.preserve_format)
+                    if amsgrad:
+                        # Maintains max of all exp. moving avg. of sq. grad. values
+                        state['max_exp_avg_sq'] = torch.zeros_like(p, memory_format=torch.preserve_format)
+
+                exp_avg.append(state['exp_avg'])
+                exp_avg_sq.append(state['exp_avg_sq'])
+
+                if amsgrad:
+                    max_exp_avg_sq.append(state['max_exp_avg_sq'])
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/adamw.py:106]: same

File: torch/optim/_multi_tensor/adamw.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                # State initialization
+                if len(state) == 0:
+                    state['step'] = 0
+                    # Exponential moving average of gradient values
+                    state['exp_avg'] = torch.zeros_like(p, memory_format=torch.preserve_format)
+                    # Exponential moving average of squared gradient values
+                    state['exp_avg_sq'] = torch.zeros_like(p, memory_format=torch.preserve_format)
+                    if amsgrad:
+                        # Maintains max of all exp. moving avg. of sq. grad. values
+                        state['max_exp_avg_sq'] = torch.zeros_like(p, memory_format=torch.preserve_format)
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/adamw.py:100]: same

File: torch/optim/_multi_tensor/adamw.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            #
+            torch._foreach_mul_(exp_avg, beta1)
+            torch._foreach_add_(exp_avg, grads, alpha=1 - beta1)
+
+            torch._foreach_mul_(exp_avg_sq, beta2)
+            torch._foreach_addcmul_(exp_avg_sq, grads, grads, 1 - beta2)
+
+            if amsgrad:
+                # Maintains the maximum of all 2nd moment running avg. till now
+                max_exp_avg_sq = [torch.max(a, b) for a, b in zip(max_exp_avg_sq, exp_avg_sq)]
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/adamw.py:127]: same

File: torch/optim/_multi_tensor/adamw.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                # Use the max. for normalizing running avg. of gradient
+                max_exp_avg_sq_sqrt = torch._foreach_sqrt(max_exp_avg_sq)
+                bias_correction_sqrt = [math.sqrt(bc) for bc in bias_correction2]
+                torch._foreach_div_scalar_list_(max_exp_avg_sq_sqrt, bias_correction_sqrt)
+                denom = torch._foreach_add(max_exp_avg_sq_sqrt, group['eps'])
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/adamw.py:132]: same

File: torch/optim/_multi_tensor/sgd.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+            if grads == []:
+                return loss
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/sgd.py:112]: should this branch be tested?

File: torch/optim/_multi_tensor/sgd.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+            if weight_decay != 0:
+                grads = torch._foreach_add(grads, params_with_grad, alpha=weight_decay)
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/sgd.py:115]: same

File: torch/optim/_multi_tensor/sgd.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+                    if 'momentum_buffer' not in states[i]:
+                        buf = states[i]['momentum_buffer'] = torch.clone(grads[i]).detach()
+                    else:
+                        buf = states[i]['momentum_buffer']
+                        buf.mul_(momentum).add_(grads[i], alpha=1 - dampening)
+
+                    bufs.append(buf)
+
+                if nesterov:
+                    torch._foreach_add_(grads, bufs, alpha=momentum)
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/sgd.py:129]: same

**Comment by vincentqb** [comment]
> Original PR #45299. The present PR fixes minor bugs that caused revert.

What's the patch you applied in the end?

File: torch/optim/_multi_tensor/adadelta.py (hunk 4 above)
Comment (izdeby) [reply torch/optim/_multi_tensor/adadelta.py:81]: done

File: torch/optim/_multi_tensor/adamax.py (hunk 5 above)
Comment (izdeby) [reply torch/optim/_multi_tensor/adamax.py:85]: done

File: torch/optim/_multi_tensor/adamw.py (hunk 6 above)
Comment (izdeby) [reply torch/optim/_multi_tensor/adamw.py:106]: done

File: torch/optim/_multi_tensor/adamw.py (hunk 7 above)
Comment (izdeby) [reply torch/optim/_multi_tensor/adamw.py:100]: done

File: torch/optim/_multi_tensor/adamw.py (hunk 8 above)
Comment (izdeby) [reply torch/optim/_multi_tensor/adamw.py:127]: Done

**Comment by izdeby** [comment]
> > Original PR #45299. The present PR fixes minor bugs that caused revert.
> 
> What's the patch you applied in the end? Is there a test to ensure this is not happening?

All that was tested [via](https://github.com/pytorch/pytorch/pull/45408/commits/3982f300aaaf1a50c59336c9f33084b025512285)

Answering your question, i have additionally tested each optimizers by running https://github.com/pytorch/vision/blob/master/references/classification/train.py  these params: `clear && python references/classification/train.py -j 2 -b 3 --epochs 2 --device cuda `, which resulted in this [update](https://github.com/pytorch/pytorch/pull/45408/commits/cfb3147e7c77fd3f89164e2c4bac08f05c807b9d).

**Review by vincentqb — APPROVED** [review]
LGTM. Given that (1) this implementation is already faster than the original one, (2) the original PR was approved, (3) the issue was fixed, (4) I'm ok with letting the two performance opportunities highlighted by @mcarilli be in a follow-up PR, I'll approve this PR, so you can decide how you want to proceed @izdeby.

File: torch/optim/_multi_tensor/sgd.py (hunk 2 above)
Comment (mcarilli) [reply torch/optim/_multi_tensor/sgd.py]: As @ngimel pointed out, many optimizers have lingering per-tensor loops that should be repackaged as MTA.  This can be done over time.  It would be great for Adam and SGD, at least, to fully avoid cpu-bound loops over tensors in 1.7.

**Comment by ngimel** [comment]
My suggestion for benchmarking is comparing mta optimizers with regular optimizers operating on a single large parameter (that results from catting all the network parameters). Properly implemented mta optimizers should be pretty close to that, slower, but not by much. Such benchmarks can alert us to lingering loops in the mta implementation.

**Comment by vincentqb** [comment]
> My suggestion for benchmarking is comparing mta optimizers with regular optimizers operating on a single large parameter (that results from catting all the network parameters). Properly implemented mta optimizers should be pretty close to that, slower, but not by much. Such benchmarks can alert us to lingering loops in the mta implementation.

oops, did I misunderstand the context of the numbers mentioned in the description above? i was under the impression this is what they covered. maybe you meant with even larger model? @izdeby @ngimel

**Comment by rgommers** [comment]
Hi @izdeby, I noticed this PR introduced a new set of `.pyi` stub files. This was undesired, we're almost done getting rid of stub files and inlining type annotations in `.py` files. See https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch. The `torch.optim` ones are a few of the last ones left, and they're about to be removed in gh-43089.

Would you mind moving these new annotations inline?

**Comment by izdeby** [comment]
> Hi @izdeby, I noticed this PR introduced a new set of `.pyi` stub files. This was undesired, we're almost done getting rid of stub files and inlining type annotations in `.py` files. See https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch. The `torch.optim` ones are a few of the last ones left, and they're about to be removed in [gh-43089](https://github.com/pytorch/pytorch/pull/43089).
> 
> Would you mind moving these new annotations inline?

@rgommers, will do!
